secure_tunneling/source/SecureTunnel.cpp

```diff
@@ -0,0 +1,146 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/iotsecuretunneling/SecureTunnel.h>
+
+namespace Aws
+{
+    namespace Iotsecuretunneling
+    {
+        DeviceApiHandle::DeviceApiHandle(Crt::Allocator *allocator) noexcept { aws_iotdevice_library_init(allocator); }
+
+        DeviceApiHandle::~DeviceApiHandle() { aws_iotdevice_library_clean_up(); }
+
+        SecureTunnel::SecureTunnel(
+            Crt::Allocator *allocator,
+            Aws::Crt::Io::ClientBootstrap *clientBootstrap,
+            Aws::Crt::Io::SocketOptions *socketOptions,
+
+            std::string accessToken,
+            aws_secure_tunneling_local_proxy_mode localProxyMode,
+            std::string endpointHost,
+
+            OnSecureTunnelingConnectionComplete onConnectionComplete,
+            OnSecureTunnelingSendDataComplete onSendDataComplete,
+            OnSecureTunnelingDataReceive onDataReceive,
+            OnSecureTunnelingStreamStart onStreamStart,
+            OnSecureTunnelingStreamReset onStreamReset,
+            OnSecureTunnelingSessionReset onSessionReset)
+        {
+            // Client callbacks
+            m_OnConnectionComplete = onConnectionComplete;
```

Comment:
If you used `std::bind` here, you wouldn't need to unwrap `this` from the user data in the static functions, you could just call them.

Reply:
I tried this:
```
m_OnConnectionComplete = std::bind(onConnectionComplete, this);
```
But my C++ is not strong enough to make this to work? In particular, what should be the type of `m_OnConnectionComplete` when I declare it in the header file? 

I also wanted to point out that `m_OnConnectionComplete` is a function pointer to a client supplied static function. When I call `m_OnConnectionComplete`
```
            auto *secureTunnel = static_cast<SecureTunnel *>(user_data);
            secureTunnel->m_OnConnectionComplete();
```
It looks like I am calling a member function of `SecureTunnel`, but it is not. This is calling a function pointer (to a client supplied function) which is a data member of `SecureTunnel`.

Reply:
Oh, good call, nevermind then. You're right.